File: BitMasking_Problems/CountTheNumberOfConsistentStrings.cpp

// Problem link : https://leetcode.com/problems/count-the-number-of-consistent-strings/description/?envType=problem-list-v2&envId=bit-manipulation

#include<bits/stdc++.h>
using namespace std;

int countConsistentStrings(string allowed, vector<string>& words) {
    // convert 'allowed' string into a Set for efficient lookup of allowed characters
    unordered_set<char> allowedSet(allowed.begin(), allowed.end());

    int consistentCount = 0;

    // iterate over each word in the 'words' array
    for(string word : words){
        bool isConsistent = true;
        // check each character in the word
        for(char ch : word){
            if(allowedSet.find(ch) == allowedSet.end()){
                // if character is not allowed, mark the word as inconsistent
                isConsistent = false;
                break;
            }
        }

        // if the word is consistent, increment the counter
        if(isConsistent){
            consistentCount++;
        }
    }
    return consistentCount;
}

int main(){
    // string allowed = "ab";
    // vector<string> words = {"ad","bd","aaab","baa","badab"};

    string allowed = "abc";
    vector<string> words = {"a","b","c","ab","ac","bc","abc"};
    cout<<countConsistentStrings(allowed, words);
    return 0;
}

File: BitMasking_Problems/CountTheNumberOfConsistentStrings.py

# Problem link : https://leetcode.com/problems/count-the-number-of-consistent-strings/description/?envType=problem-list-v2&envId=bit-manipulation

from typing import List

def countConsistentStrings(allowed: str, words: List[str]) -> int:
    # convert 'allowed' string into a Set for efficient lookup of allowed characters
    allowed_set = set(allowed)

    consistent_count = 0

    # iterate over each word in the 'words' array
    for word in words:
        # check each character in the word
        if all(c in allowed_set for c in word):
            # if consistent, increment the counter
            consistent_count += 1

    # return the final count  
    return consistent_count

# allowed = "ab"
# words = ["ad","bd","aaab","baa","badab"]

allowed = "abc"
words = ["a","b","c","ab","ac","bc","abc"]
print(countConsistentStrings(allowed, words))